prefect_azure/container_instance.py

```diff
@@ -0,0 +1,694 @@
+    ContainerInstanceJob(command=["echo", "hello world"], image="alpine:latest")
+    ```
+
+    Run a task with custom memory and CPU requirements
+    ```python
+    ContainerInstanceJob(command=["echo", "hello world"], memory=1.0, cpu=1.0)
+    ```
+
+    Run a task with custom memory and CPU requirements
+    ```python
+    ContainerInstanceJob(command=["echo", "hello world"], memory=1.0, cpu=1.0)
+    ```
+
+    Run a task with custom memory, CPU, and GPU requirements
+    ```python
+    ContainerInstanceJob(command=["echo", "hello world"], memory=1.0, cpu=1.0,
+    gpu_count=1, gpu_sku="V100")
+    ```
+
+    Run a task with custom environment variables
+    ```python
+    ContainerInstanceJob(command=["echo", "hello $PLANET"], env={"PLANET": "earth"})
+    ```
+"""
+import datetime
+import json
+import sys
+import time
+import uuid
+from enum import Enum
+from typing import Dict, List, Optional
+
+import dateutil.parser
+import prefect.infrastructure.docker
+from anyio.abc import TaskStatus
+from azure.core.credentials import TokenCredential
+from azure.core.polling import LROPoller
+from azure.identity import ClientSecretCredential
+from azure.mgmt.containerinstance import ContainerInstanceManagementClient
+from azure.mgmt.containerinstance.models import (
+    Container,
+    ContainerGroup,
+    ContainerGroupRestartPolicy,
+    ContainerGroupSubnetId,
+    EnvironmentVariable,
+    GpuResource,
+    ImageRegistryCredential,
+    OperatingSystemTypes,
+    ResourceRequests,
+    ResourceRequirements,
+)
+from azure.mgmt.resource import ResourceManagementClient
+from prefect.docker import get_prefect_image_name
+from prefect.infrastructure.base import Infrastructure, InfrastructureResult
+from prefect.utilities.asyncutils import run_sync_in_worker_thread, sync_compatible
+from pydantic import Field, SecretStr
+from typing_extensions import Literal
+
+from prefect_azure.credentials import ContainerInstanceCredentials
+
+ACI_DEFAULT_CPU = 1.0
+ACI_DEFAULT_MEMORY = 1.0
+ACI_DEFAULT_GPU = 0.0
+DEFAULT_CONTAINER_ENTRYPOINT = "/opt/prefect/entrypoint.sh"
+
+
+class ContainerGroupProvisioningState(str, Enum):
+    """
+    Terminal provisioning states for ACI container groups. Per the Azure docs,
+    the states in this Enum are the only ones that can be relied on as dependencies.
+    """
+
+    SUCCEEDED = "Succeeded"
+    FAILED = "Failed"
+
+
+class ContainerRunState(str, Enum):
+    """
+    Terminal run states for ACI containers.
+    """
+
+    RUNNING = "Running"
+    TERMINATED = "Terminated"
+
+
+class ContainerInstanceJobResult(InfrastructureResult):
+    """
+    The result of an `ContainerInstanceJob` run.
+    """
+
+
+class ContainerInstanceJob(Infrastructure):
+    """
+    <span class="badge-api experimental"/>
+    Run a command using a container on Azure Container Instances.
+    Note this block is experimental. The interface may change without notice.
+    """
+
+    _block_type_slug = "container-instance-job"
```

Comment:
Anna mentioned that it would be useful for the block type slug to start with 'azure' since the slug is used when creating CLI deployments and "container-instance-job" sounds a bit generic. Not our fault; Azure's branding of this service just isn't very descriptive. 

I don't think it's a blocker, just a potential user QOL enhancement; most people who install this collection to use ACI infrastructure should also be able to figure things out with the current block type slug.

Reply:
Ah block type slugs need to be globally unique and can't change after release. We'll definitely want this to begin with `azure-`

Reply:
If you remove this line, the slug will automatically become `azure-container-instance-job`.